Feature/format to string

juliusnehring opened the pull request:
Please note that the implementation is very rudimentary and also very much work in progress. Please review primarily for API, and decide whether such WIP implementation is ok.

juliusnehring committed:
First working format version.

juliusnehring committed:
Removed unnecessary verify_args function

juliusnehring committed:
Can now use a formatter other than the default formatter.

juliusnehring committed:
Cleanup. Much of the logic moved to to_string

juliusnehring committed:
Cleanup

juliusnehring committed:
WIP

juliusnehring committed:
WIP

juliusnehring committed:
Cleanup of exposed api.
Fix: end checked before value.
usage of char_predicates.hh.

juliusnehring committed:
Merge branch 'develop' into feature/format_to_string

juliusnehring committed:
Fixed some safety bound checks.

juliusnehring committed:
Bugfix: empty id with format arguments lead to a crash.

juliusnehring committed:
Added base implementation for formatted to_string(...)

Philip-Trettner commented on src/clean-core/to_string.hh:
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <clean-core/string.hh>
+#include <clean-core/string_stream.hh>
```
now that this header is dual-purpose we should only forward declare those two

juliusnehring replied:
Done. Moved remaining definitions into the `to_string.cc`

juliusnehring committed:
Moved definitions into .cc to allow forward declaration of string and string_view

juliusnehring committed:
Added `string to_string(value, fmt_str)` and `void to_string(string_stream&, value)`

Philip-Trettner commented on src/clean-core/format.hh:
```diff
@@ -0,0 +1,167 @@
+namespace cc
+{
+template <class T>
+struct arg;
```
why is this in the "global" `cc` namespace?

juliusnehring replied:
This is how you define named arguments. ` cc::format("{h} {w}!", cc::arg("w", "World"), cc::arg("h", "Hello"));`

juliusnehring replied:
`cc::arg` was renamed to `cc::format_arg` and a user defined literal was added for better ease of use.

juliusnehring committed:
Changed cc::arg to cc::format_arg
Changed cc::format_arg name member from char const* to string_view
Added user defined literal _a for named arguments.

jkunstwald commented on src/clean-core/format.hh:
```diff
@@ -0,0 +1,184 @@
+}
+
+void vformat_to(string_stream& ss, string_view fmt_str, span<arg_info> args);
+}
```
Maybe we should start using the `} // end namespace X` convention for things like this

juliusnehring replied:
I am absolutely in favor of that!

jkunstwald commented on src/clean-core/format.hh:
```diff
@@ -0,0 +1,184 @@
+};
+
+template <class Formatter = detail::default_formatter, class... Args>
+void format_to(string_stream& ss, string_view fmt_str, Args const&... args)
```
I'd like a comment explaining usage, it's not quite self evident :D

juliusnehring replied:
The basic use-case is of course similar to `cc::format("{}", "Hello World!");`, but instead we format into a `cc::string_stream`. 
```cpp
cc::string_stream ss;
cc::format_to(ss, "{} {}!", "Hello", "World");
```
However, you are probably referring to the `Formatter` template argument :)
The reason this exists is to allow customization of formatting any type, including the builtin types.
Usually, types are formatted by their corresponding `to_string(....)` function.
However, this might not always be what you want for your use-case.
Consider a formatter that works as before for all types except for `int`, but instead of a numerical representation inserts the character sequence "int".
```cpp
struct my_formatter
{
    template <class T>
    static void do_format(string_stream& ss, T const& v, string_view fmt_args)
    {
         cc::default_formatter::do_format(ss, v, fmt_args);
    }
    static void do_format(string_stream& ss, int const&, string_view)
    {
        ss << "int";
    }
};
```
Usage is then as follows:

```cpp
cc::string_stream ss;
cc::format_to<my_formatter>(ss, "{}", 42);
```
The result is now `"int"`.

juliusnehring replied:
And if your question was about syntax for the `fmt_string`, it is essentially the same as for [fmtlib](https://fmt.dev/latest/). 

It is planned to extend the syntax slightly to enable all types (even custom ones) to use alignment options without having to re-implement the logic in their `to_string(...)`. Currently the part inside `{...}` consists of an optional id, which can be followed by `:` and some type specific options (have a look at fmtlibs [syntax page](https://fmt.dev/latest/syntax.html) for more information). The plan is to allow a second `:`, after which an additional set of alignment options can be declared.
Example: `{my_arg_name:my_type_specific_options:-<100}`.

jkunstwald replied:
I mostly meant a comment in code :) But thanks!

jkunstwald commented on src/clean-core/format.hh:
```diff
@@ -0,0 +1,184 @@
+};
+template <class T>
+struct has_to_string_ss_t<T, std::void_t<decltype(to_string(std::declval<string_stream>(), std::declval<T>()))>> : std::true_type
+{
+};
+template <class T>
+constexpr bool has_to_string_ss = has_to_string_ss_t<T>::value;
+
+template <class T, class = std::void_t<>>
+struct has_to_string_args_t : std::false_type
+{
+};
+template <class T>
+struct has_to_string_args_t<T, std::void_t<decltype(string_view(o_string(std::declval<T>(), std::declval<string_view>())))>> : std::true_type
+{
+};
+template <class T>
+constexpr bool has_to_string_args = has_to_string_ss_t<T>::value;
+
+template <class T, class = std::void_t<>>
+struct has_to_string_t : std::false_type
+{
+};
+template <class T>
+struct has_to_string_t<T, std::void_t<decltype(string_view(to_string(std::declval<T>())))>> : std::true_type
+{
+};
+template <class T>
+constexpr bool has_to_string = has_to_string_t<T>::value;
+
+template <class T, class = std::void_t<>>
+struct has_member_to_string_t : std::false_type
+{
+};
+template <class T>
+struct has_member_to_string_t<T, std::void_t<decltype(string_view(std::declval<T>().to_string()))>> : std::true_type
+{
+};
+template <class T>
+constexpr bool has_member_to_string = has_member_to_string_t<T>::value;
```
Maybe these helpers could be moved to a separate header in detail/

juliusnehring replied:
That would certainly make the header much cleaner. Do you have a naming suggestion?

jkunstwald replied:
Mabye `has_to_string.hh`? Then again `reference_wrapper.hh` doesn't include the `is_` prefix

jkunstwald commented on src/clean-core/format.hh:
```diff
@@ -0,0 +1,184 @@
+            &a.value, a.name};
+}
+
+void vformat_to(string_stream& ss, string_view fmt_str, span<arg_info> args);
```
Should this be `span<arg_info const>`?

juliusnehring replied:
Good point!

jkunstwald commented on src/clean-core/to_string.hh:
```diff
@@ -1,27 +1,106 @@
-inline string to_string(char const* value) { return value == nullptr ? "[nullptr]" : value; }
-inline string to_string(string_view value) { return value; }
-inline string to_string(nullptr_t) { return "nullptr"; }
+string to_string(char value);
+string to_string(bool value);
+string to_string(char const* value);
+string to_string(string_view value);
+string to_string(nullptr_t);
```
Im not sure whether not including string here is worth not being able to inline these calls. Technically all but the first one could just return `char const*` anyway

Philip-Trettner replied:
I've discussed this with julius and the voted for this solution because calling to_string in a hot loop is not a good idea anways.
I don't like `char const*` return value because that would be wholly inconsistent (why is `to_string(2) + " foo"` working but not `to_string(true) + " foo"`? sounds like a big headache).
btw, `string_view` is not null-terminated so `char const*` would not work.

jkunstwald replied:
I see, i was just thinking since the format internals rely so heavily on these functions and currently live in a different TU. Completely agree on the consistency part.

juliusnehring replied:
Aren't all `cc` internals currently in a single TU?

jkunstwald replied:
Right!

jkunstwald commented on src/clean-core/to_string.cc:
```diff
@@ -95,3 +141,834 @@ cc::string cc::to_string(std::byte value)
+{
+    // note: currently completely relys on std::sprintf
+
+    cc::string_stream sprintf_args;
```
Seems to me that char capped_vector or similar could do this job more efficiently

Philip-Trettner replied:
it's not _that_ easy because later a `to_string(sprintf_args, args.width, "");` is used

juliusnehring replied:
Also I plan to get rid of all calls to `std::sprintf`, so the implementation is bound to change quite significantly :)

jkunstwald replied:
Yes, the API really would benefit from stream_ref here (although this particular case could just use another snprintf)

jkunstwald commented on src/clean-core/to_string.cc:
```diff
@@ -95,3 +141,834 @@ cc::string cc::to_string(std::byte value)
+        sprintf_args << "G";
+        break;
+    default:
+        CC_ASSERT(cc::always_false<FloatType> && "Invalid format string: Unsupported argument type for float type");
```
Runtime asserts should simply be `false`

jkunstwald commented on src/clean-core/to_string.cc:
```diff
@@ -95,3 +141,834 @@ cc::string cc::to_string(std::byte value)
+    char buffer[buffer_size];
+    auto const s = sprintf_args.to_string();
+    auto res = std::snprintf(buffer, sizeof(buffer), sprintf_args.to_string().c_str(), value);
+    CC_ASSERT(res >= 0);
```
The error catching switch above could be replaced by passing through `args.type` and testing for snprintf errors like this. For sanity regarding the 31/41 above, this could also assert that `res <= buffer_size`

juliusnehring replied:
I actually just realized that this is not correct anyways. The buffer size is 31 for `double`, 41 for `long double`.
I am not sure I understand what you mean by passing `args.type`. It does not distinguish `float` from `double`.
I don't think testing` res <= buffer_size` is necessary, as `snprintf` is already passed the buffer size.

jkunstwald replied:
If snprintf returns a number > buffer size, it would have "wanted" to write more chars than it did, ie. the stack buffer wasn't large enough.

jkunstwald replied:
With passing args.type i meant just doing `sprintf_args << args.type` without the switch, as this assert would catch formatting errors as well

juliusnehring replied:
Ah ok 👍

juliusnehring replied:
Ya I already warned Philip about that switch statement, it's quite useless at the moment :D . I was planning to replace its inner parts by calls to `int_to_hex(...)`, `int_to_octa(...)`, ect.  but rather rushed a `snprintf` based solution to get a working solution. The focus for an initial version should really be API.

juliusnehring replied:
Wrong switch statement (not the integer one), but same concept.

jkunstwald reviewed (commented):
Really appreciate the amount of work gone into this. 
As with #75, im not sure about some core choices in this architecture, for example the fact that it's impossible to directly format into an arbitrary buffer, the only possible output is via an allocating container.
In my mind we'd have to beat fmtlib, C++20 fmt, and C-style printf in at least one of compile-time perf, runtime perf or usability and safety, and im not sure if that's the case right now.

Philip-Trettner commented:
> As with #75, im not sure about some core choices in this architecture, for example the fact that it's impossible to directly format into an arbitrary buffer, the only possible output is via an allocating container.

we could change that to `stream_ref<char>` once I implemented that.

> In my mind we'd have to beat fmtlib, C++20 fmt, and C-style printf in at least one of compile-time perf, runtime perf or usability and safety, and im not sure if that's the case right now.

we definitely beat fmtlib and fmt in compile-time perf.
runtime-perf is something we can improve behind the scenes as long as the API allows it (which I think it does).
usability and safety is on par with fmt and obv superior to printf (due to proper string interaction).
the follow-up `rf::format` improves usability even further.

note that compile-time perf is a major topic as fmt alone can bring down you builds when used in headers.

jkunstwald commented:
@Philip-Trettner 
> we could change that to stream_ref<char> once I implemented that.

I think this PR should wait until that is done, since it would be a rather wide ranging breaking API change regarding the user-implemented to_strings

> runtime-perf is something we can improve behind the scenes as long as the API allows it (which I think it does).

I wonder how fmt is faster than printf in the first place, i assume it's related to constexpr parsing of the format string in some way? 

> usability and safety is on par with fmt and obv superior to printf (due to proper string interaction).

I'm not so sure about this part, both fmt and printf have compile time errors regarding the format string, we just have runtime asserts

> we definitely beat fmtlib and fmt in compile-time perf.

That's very good. 

As for API in general, i'd like a version with an interface similar to snprintf, ie. 
```C++
int format(span<char> out_buffer, char const* fmt, ...);
```
which would differ from the `stream_ref` one by accepting limited-size out buffers (and ie. would be usable in the situation above)

juliusnehring commented:
@jkunstwald 
The constexpr parsing is very limited in both cases. For printf you have some specific non-standard compiler warnings afaik, and for fmtlib you do not get compile time checks by default, but instead you need to wrap your char literal in a macro and also pay the price of limited functionality as a result.

I think the main reason why printf is so slow is, because it is locale aware.

jkunstwald commented:
@juliusnehring 
> for fmtlib you do not get compile time checks by default, but instead you need to wrap your char literal in a macro and also pay the price of limited functionality as a result.

Ah, i see. 
With the `stream_ref` change in place i think this PR is good to go.

(Just a note, wouldn't call printf slow, fmt takes 0.75 seconds while printf takes 1.04 
 in their self-devised benchmark on O3: https://github.com/fmtlib/fmt#speed-tests - it's probably quite a task to reach even just printf perf, and safe to say that fmt is leagues behind on debug)

juliusnehring commented:
As an alternative to a solution that requires `stream_ref<T>`, we could also move the implementation into a templated header, but provide a default non-templated implementation that uses `string_stream` and only includes the templated header inside the format.cc file.

jkunstwald commented:
The stream ref would be strictly char, not T, and string_stream converts to it

juliusnehring commented:
Yeah the T got there on accident, `cc::format` is only char based as well :)

What i meant is a version that is templated on the stream and can be fully inlined into user code.

juliusnehring committed:
Changed arguments from span<arg_info> to span<arg_info const>

juliusnehring committed:
Removed unnecessary switch statement
added snprintf result sanity check for buffer size.

juliusnehring committed:
Merge branch 'udevelop' into feature/format_to_string

juliusnehring committed:
Replaced string_stream with stream_ref<char>
